include/circt/Support/PrettyPrinter.h

```diff
@@ -72,14 +72,24 @@ class Token {
   struct EndInfo : public TokenInfo {
     // Nothing
   };
+  // This can be used to associate a callback with the print event on the
+  // tokens stream. The id, is used to uniquely identify an intereseting event
+  // on the token stream.
+  struct CallbackInfo : public TokenInfo {
+    using CallbackTy = llvm::function_ref<void(unsigned)>;
+    CallbackTy callback;
```

Comment:
The function_ref is already the size of the other tokens (16), this increases the token sizes 16 -> 32 (on my machine anyway).

Maybe the callback should just be on the listener? Storing function AND identifier seems overkill.